### PR TITLE
Implement AuthN against OpenStack Keystone using password method

### PR DESCRIPTION
**Pull request by gbraad**
Basic implementation of password authentication against OpenStack Keystone #21

### CODE REVIEW

**Comment by gbraad** [comment]
Would like to implement testcases to mock the keystone endpoint, just not sure what would be the best approach.

**Comment by ashcrow** [comment]
I'll lend a hand with coming up with a mock pattern.

**Comment by gbraad** [comment]
I suggest to work on the token authentication in a separate Pull Request.

File: src/commissaire_http/authentication/__init__.py [hunk 1]
```diff
@@ -67,3 +68,30 @@ def authenticate(self, environ, start_response):
...
+    :returns: tuple -- (username, passphrase) or (None, None) if empty.
+    :rtype: tuple
+    """
+    if http_auth is not None:
+        if http_auth.lower().startswith('basic '):
+            try:
+                decoded = tuple(base64.decodebytes(
+                    http_auth[6:].encode('utf-8')).decode().split(':'))
+                if logger:
+                    logger.debug('Credentials given: {0}'.format(decoded))
```
Comment (ashcrow) [inline src/commissaire_http/authentication/__init__.py]: Almost certainly from our original code but just a reminder that we no longer need to use ``{N}`` with Python 3. It won't hurt anything to use it and it's not worth changing but, when new code is written using ``{}`` is considered preferable.

File: src/commissaire_http/authentication/__init__.py [hunk 2]
```diff
@@ -67,3 +68,30 @@ def authenticate(self, environ, start_response):
         :rtype: bool
         """
         return False
+
+
+def decode_basic_auth(logger, http_auth):
```
Comment (ashcrow) [inline src/commissaire_http/authentication/__init__.py]: :+1: for breaking this out!

File: src/commissaire_http/authentication/keystonepasswordauth.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        Checks the user name and password from an Authorization header
+        against the specified OpenStack Keystone instance
+
+        :param app: The WSGI application being wrapped with authenticaiton.
+        :type app: callable
+        :param url: The OpenStack Keystone endpoint used for Authentication
+        :type url: string
+        :returns: HTTPBasicAuth
+        """
+        super(KeystonePassword, self).__init__(app)
```
Comment (ashcrow) [inline src/commissaire_http/authentication/keystonepasswordauth.py]: Another nice Python 3 feature: ``super`` doesn't require input unless there is something special needed. This line could read: ``super().__init__(app)``. Not worth changing as this works totally fine just want to spread the Python 3 happiness :smile:.

**Comment by ashcrow** [comment]
bot, test pull request

File: src/commissaire_http/authentication/__init__.py (hunk 1 above)
Comment (gbraad) [inline src/commissaire_http/authentication/__init__.py]: Thanks. Wasn't aware of this

**Comment by gbraad** [comment]
Thanks for the feedback. this is why I pushed for PR as soon as possible. PR-first makes people aware a task is being worked on, but also allows for early feedback.

**Pull request merged**